### PR TITLE
Fix multiple widget issues and improve daily audit logging

### PR DESCRIPTION
**Pull request by OPS-PIvers**
This pull request introduces several improvements and bug fixes across the codebase, focusing on widget behavior, accessibility, and event handling. The most significant changes address a regression in the NextUp widget's auto-expiry logic, improve the reliability and responsiveness of the SoundWidget's visualizer, and enhance keyboard event handling to prevent unintended UI interactions. Additionally, accessibility improvements have been made to the SyntaxFramer widget's settings.

**Widget Logic and Bug Fixes:**

* The NextUp widget now properly auto-expires sessions that span across midnight, even if no other state changes occur, by introducing a periodic timer (`nowTick`) and extracting the date comparison logic to a pure function (`shouldExpireNextUpQueue`). This resolves an issue where idle sessions could remain active indefinitely. [[1]](diffhunk://#diff-309adad4eb1a8e89c3951854df8fc34ecff01dbe69105f36c4e2e49ff210302eR51-R83) [[2]](diffhunk://#diff-1fd4e260bd24e8e46026e9af2485eaf1188146cc83800963b94ad51351ff43d9R24-R43) [[3]](diffhunk://#diff-8b7a1405b11a6a613a55ce22a282d80cc89415125b2930c56bc13e10e5a54ff4R123-R183) [[4]](diffhunk://#diff-968c9ff49d1f626d30cc78edc2512eed96de48a89bd9b00e60fb61acd8aedc5eR1-R121) [[5]](diffhunk://#diff-8b7a1405b11a6a613a55ce22a282d80cc89415125b2930c56bc13e10e5a54ff4L19-R22)
* Added comprehensive regression tests to ensure the NextUp widget's auto-expiry works as intended, covering edge cases such as overnight inactivity and missing timestamps. [[1]](diffhunk://#diff-968c9ff49d1f626d30cc78edc2512eed96de48a89bd9b00e60fb61acd8aedc5eR1-R121) [[2]](diffhunk://#diff-8b7a1405b11a6a613a55ce22a282d80cc89415125b2930c56bc13e10e5a54ff4R123-R183)
* The SoundWidget's "balls" visualizer (`PopcornBallsView`) now dynamically measures its container using a `ResizeObserver`, ensuring the canvas always matches the rendered area and removing the need for hard-coded pixel adjustments. [[1]](diffhunk://#diff-6b651a4370fd03bc0244db0ee55d42f8361710c5bb207fe0f0a2c914ebe22ea9L1-R17) [[2]](diffhunk://#diff-6b651a4370fd03bc0244db0ee55d42f8361710c5bb207fe0f0a2c914ebe22ea9R28-R46) [[3]](diffhunk://#diff-6b651a4370fd03bc0244db0ee55d42f8361710c5bb207fe0f0a2c914ebe22ea9R60) [[4]](diffhunk://#diff-6b651a4370fd03bc0244db0ee55d42f8361710c5bb207fe0f0a2c914ebe22ea9R96-R103) [[5]](diffhunk://#diff-c50bea013655cd0551177c86515a4995787a9e06681eec9bfb2e9ec01330a93bL300-R298)

**Event Handling Improvements:**

* Updated popover and menu components (`ToolDockItem`, `ClassRosterMenu`, `RemoteControlMenu`) to call `event.stopPropagation()` on Escape key events. This prevents Escape from bubbling up and triggering unrelated global handlers, such as minimizing widgets unintentionally. [[1]](diffhunk://#diff-5e5fe55075be14b0823390a194f9ac141b40411dc9aca561ca01f65a04196e71R88-R110) [[2]](diffhunk://#diff-d32076a12a65d9a712f4d587616790fc9637561858080d1df8af20ac4c543a76R51-R56) [[3]](diffhunk://#diff-513d1e916507ef962e369bec85947a20b114a5b7614ec9b1ae307007e6cad77aR72-R77)

**Accessibility Enhancements:**

* Improved accessibility for the SyntaxFramer widget's alignment controls by associating the button group with an accessible label using `aria-labelledby` and unique IDs. [[1]](diffhunk://#diff-c5bf6076c611c079bb904744824d964255fc5084a4fbea6010496654ded1d91dR155) [[2]](diffhunk://#diff-c5bf6076c611c079bb904744824d964255fc5084a4fbea6010496654ded1d91dL165-R177)

**Other Minor Changes:**

* Standardized error message apostrophe usage in `InviteAcceptance.tsx` for consistency.
* Cleaned up unused variables in `SoundWidget/Widget.tsx`.

These changes collectively improve the reliability, usability, and accessibility of the application.

### CODE REVIEW

**Comment by OPS-PIvers** [comment]
Automated review pass — no unresolved threads found; one new observation not covered by the prior review:

**`ToolDockItem.tsx` — focus restoration after Escape is untested**

The new Escape handler in `ToolDockItem.tsx` calls `buttonRef.current?.focus()` to return keyboard focus to the dock button after the popover closes:

```tsx
e.stopPropagation();
setShowPopover(false);
buttonRef.current?.focus();
```

`tests/components/layout/ToolDockItem.test.tsx` verifies both that the popover closes and that the keydown doesn't leak to `window`, but doesn't assert that focus actually lands back on the button. Because the call uses optional chaining (`?.`), a case where `buttonRef.current` is null would silently skip focus restoration with no test failure — important to catch since keyboard-only teachers depend on this for sequential dock navigation.

A minimal assertion to add to the second test (or a dedicated third test):

```tsx
const button = screen.getByRole('button', { name: /timer/i });
fireEvent.keyDown(document, { key: 'Escape', bubbles: true });
expect(document.activeElement).toBe(button);
```

The sibling `ClassRosterMenu` and `RemoteControlMenu` tests don't test focus restoration either, but those components don't explicitly call `.focus()` in their handlers — `ToolDockItem` is the only one that does, which is why this gap is specific to this file.

Everything else looks good — the `stopPropagation` fix, the `ResizeObserver` self-measurement, the `shouldExpireNextUpQueue` extraction + 60 s ticker, the pagination cursor fix, and the `aria-labelledby` retrofit are all correct.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by OPS-PIvers** [comment]
Addressed both actionable review observations in `c9d185e`:

**1. `ToolDockItem.tsx` — focus restoration after Escape is untested** (from the automated review pass above)

Valid gap. The handler calls `buttonRef.current?.focus()`, but no test asserted focus actually lands back on the dock button — the optional chain would silently skip restoration if `buttonRef.current` were ever null, with no failure. Added a third test to `tests/components/layout/ToolDockItem.test.tsx` that presses Escape and asserts `document.activeElement` is the dock button. (Targeted the dock trigger by its `data-tool-id` since the popover's restore entry shares the "Timer" accessible name.) All three tests pass.

**2. NextUp auto-expiry effect deps** (minor nit in the `claude[bot]` review)

Valid, harmless cleanup. `config.isActive` and `config.createdAt` were both subsumed by the whole-object `config` dep (the update spreads `...config`), so listing them separately was redundant. Tightened the deps array in `components/widgets/NextUp/Widget.tsx` to `[config, widget.id, updateWidget, nowTick]` and added a short comment noting why the member reads don't need their own entries. Type-check, lint (`--max-warnings 0`), and the NextUp test suite all pass.

The remaining review notes were explicitly non-blocking observations (the `expireActivityWallShares` exact-ceiling warning is intentional conservative behaviour, and the `claude[bot]` review was otherwise LGTM), so no further changes there. The `chatgpt-codex-connector` comment is a rate-limit notice, not a review suggestion — no action.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: components/widgets/NextUp/nextUpQueueUtils.ts [hunk 1]
```diff
@@ -21,3 +21,23 @@ export function advanceNextUpQueue(
...
+ * ticking clock — the caller is responsible for re-invoking this on a
+ * periodic tick (see Widget.tsx) so day-rollover is detected even when
+ * `config` hasn't otherwise changed since before midnight.
+ */
+export function shouldExpireNextUpQueue(
+  isActive: boolean | undefined,
+  createdAt: number | undefined,
+  now: Date
+): boolean {
+  if (!isActive || !createdAt) return false;
```
[the remark this answers is not in the document]
Comment (OPS-PIvers) [reply components/widgets/NextUp/nextUpQueueUtils.ts]: Valid — applied in `f0156f1`. Changed the guard to `if (!isActive || createdAt == null) return false;`. `createdAt === 0` (the Unix epoch) is a real prior-day timestamp that should expire, and `== null` narrows the guard to the genuinely-missing case (`undefined`/`null`) as intended. The existing "missing createdAt" test uses `undefined`, so it still passes, and I added a dedicated regression test asserting `shouldExpireNextUpQueue(true, 0, now)` returns `true`. Type-check, lint, and the NextUp suites are green.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: components/widgets/SoundWidget/components/PopcornBallsView.tsx [hunk 2]
```diff
@@ -17,8 +25,25 @@ export const PopcornBallsView: React.FC<{
     volumeRef.current = volume;
   }, [volume]);
```
[the remark this answers is not in the document]
Comment (OPS-PIvers) [reply components/widgets/SoundWidget/components/PopcornBallsView.tsx:31]: Partially applied in `f0156f1`.

**Item 2 (the `⚡ Bolt Optimization` tag):** fixed — replaced with a plain comment explaining why `volume` is mirrored into a ref (so the rAF render loop, keyed only on `width`/`height`, reads the latest value without being torn down on every volume change).

**Item 1 (drop the effect, write the ref in the render body):** not applied, and worth flagging. The suggested `volumeRef.current = volume;` during render trips the repo's `react-hooks/refs` ESLint rule (`Cannot update ref during render`), which is a hard CI gate here:

```
27:3  error  Cannot access refs during render ... react-hooks/refs
```

So although CLAUDE.md's prose says "assign refs directly in the render body," the enabled lint rule forbids an unconditional ref write during render, and CI blocks on it. The effect-based sync is the lint-clean way to keep a latest-value ref, so I kept it. Lint (`--max-warnings 0`), type-check, and format-check all pass on the file.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by OPS-PIvers** [comment]
Thanks for the review. One correction on the `volumeRef` note, since it disputes a deliberate choice and a future "cleanup" based on it would break CI:

The `react-hooks/refs` rule **is** enabled in this repo (it ships with the React-Compiler-aware `eslint-plugin-react-hooks`). Running the suggested render-body write locally produces a hard error under `--max-warnings 0`:

```
components/widgets/SoundWidget/components/PopcornBallsView.tsx
  27:3  error  Cannot update ref during render ... react-hooks/refs
```

And the cited precedent actually supports keeping the effect: `sensitivityRef` in `SoundWidget/Widget.tsx` (lines 49–52) uses the **same effect-based sync**, not a render-body assignment —

```tsx
const sensitivityRef = useRef(sensitivity);
useEffect(() => {
  sensitivityRef.current = sensitivity;
}, [sensitivity]);
```

Every other `.current =` in that file is inside an effect or callback, never in render. So `PopcornBallsView` now matches the established pattern in its sibling, and the effect is the lint-clean way to keep a latest-value ref here. Leaving it as-is; the comment wording has been updated to reference the enabled lint rule rather than implying it's a universal one.

The rest (config-in-deps being harmless, the exact-5000 ceiling warning being a benign safety log) matches my read — no changes there.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: tests/components/widgets/SoundWidget/components/PopcornBallsView.test.tsx [hunk 3]
```diff
@@ -1,33 +1,73 @@
...
+    const { container, unmount } = render(<PopcornBallsView volume={50} />);
+    // No animation until the container has a non-zero measured size.
+    expect(window.requestAnimationFrame).not.toHaveBeenCalled();
+
+    const measured = container.querySelector('div');
+    emitResize(measured as Element, 400, 300);
+
     expect(window.requestAnimationFrame).toHaveBeenCalled();
     unmount();
     expect(window.cancelAnimationFrame).toHaveBeenCalled();
```
[the remark this answers is not in the document]
Comment (OPS-PIvers) [reply tests/components/widgets/SoundWidget/components/PopcornBallsView.test.tsx:78]: Valid — applied in `d4d0e25`. Pinned the teardown by asserting `disconnect` is called on unmount in the existing "starts animation once measured and stops it on unmount" test.

I adapted the mechanism: `vi.mocked(ResizeObserver).mock.instances[0]` doesn't work here because `ResizeObserver` is installed via `vi.stubGlobal('ResizeObserver', MockResizeObserver)` (a plain class, not a `vi.fn()`), so it has no `.mock.instances`. Instead the mock captures its own `disconnect` spy into a module-level `disconnectSpy` in the constructor (assigning `this.disconnect`, a property — not `this` itself, which would trip `@typescript-eslint/no-this-alias`), and the test asserts `expect(disconnectSpy).toHaveBeenCalled()`. Test passes; lint and type-check green.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: functions/src/expireActivityWallShares.ts [hunk 4]
```diff
@@ -67,12 +85,52 @@ function sessionIdOf(data: SharedActivityWallData): string | null {
...
+    if (lastDoc) query = query.startAfter(lastDoc);
+    const page = await query.get();
+    if (page.size === 0) break;
+
+    results.push(...page.docs);
+    visited += page.size;
+    lastDoc = page.docs[page.docs.length - 1];
+    if (page.size < LAPSED_SHARE_PAGE_SIZE) break;
+  }
+  if (visited >= MAX_LAPSED_SHARES_PER_RUN) {
```
[the remark this answers is not in the document]
Comment (OPS-PIvers) [reply functions/src/expireActivityWallShares.ts:118]: Valid — applied in `d4d0e25`. Added an oncall-facing comment above the ceiling warning documenting the exact-`MAX` nuance: reaching the ceiling means we stopped on the cap (not because pages ran out), so at exactly 5000 with a full final page every doc was processed but we can't prove it without another query — hence the conservative warn. No behavior change. Functions build (`tsc`), lint (`--max-warnings 0`), and format-check pass.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: components/layout/dock/ToolDockItem.tsx [hunk 5]
```diff
@@ -84,6 +85,29 @@ export const ToolDockItem = React.memo(
...
+      if (!showPopover) return undefined;
+      const handleKeyDown = (e: KeyboardEvent) => {
+        if (e.key !== 'Escape') return;
+        if (isEscapeFromWidgetInput(e)) return;
+        e.stopPropagation();
+        setShowPopover(false);
+        buttonRef.current?.focus();
+      };
+      document.addEventListener('keydown', handleKeyDown);
+      return () => document.removeEventListener('keydown', handleKeyDown);
```
[the remark this answers is not in the document]
Comment (OPS-PIvers) [reply components/layout/dock/ToolDockItem.tsx:108]: Not applied — this one would actually break CI. The premise ("exhaustive-deps knows refs are stable and won't complain if absent") holds for a *direct* `useRef()` result, but `buttonRef` here isn't one:

```tsx
const buttonRef = externalButtonRef ?? internalButtonRef;
```

It's a value computed each render from a prop (`externalButtonRef`) coalesced with a ref. `exhaustive-deps` can't prove the stability of that expression through the `??`, so it treats `buttonRef` as a required dependency. Removing it from the array produces:

```
109:8  warning  React Hook useEffect has a missing dependency: 'buttonRef'.
                 Either include it or remove the dependency array  react-hooks/exhaustive-deps
```

Under this repo's `--max-warnings 0` gate that warning fails lint and blocks CI (verified locally just now). So the entry isn't merely harmless-but-redundant here — it's required by the linter. Leaving the deps as `[showPopover, buttonRef]`.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by OPS-PIvers — COMMENTED** [review]
## Automated Code Review

**Date:** 2026-07-24

### Summary
Integration PR merging `dev-paul` into `main`. Bundles several widget fixes: NextUp auto-expiry across midnight (via a pure `shouldExpireNextUpQueue` + a periodic `nowTick`), SoundWidget `PopcornBallsView` `ResizeObserver`-based canvas measurement, Escape-key `stopPropagation` on popover/menu components, SyntaxFramer alignment-control a11y, and `expireActivityWallShares` pagination.

### Code Quality
✓ No issues found. All five inline review threads are resolved with owner verification: the `createdAt == null` epoch-0 guard fix (`f0156f1`), the ref-sync effect kept for the `react-hooks/refs` lint gate (with the non-standard `⚡ Bolt Optimization` tag removed), the pinned `ResizeObserver.disconnect` teardown assertion, the oncall ceiling-warning comment, and the justified `buttonRef` dep retention.

### Widget Standards
✓ No issues found. Changes touch existing widgets (NextUp, SoundWidget, SyntaxFramer) without altering registry entries or scaling contracts.

### Admin Config Standards
✓ Not applicable.

### Firestore & Security
✓ No issues found. `expireActivityWallShares` is a Cloud Function sweep over an existing collection; no new collections or rules, and the bounded-page ceiling (`MAX + PAGE_SIZE - 1`) is documented and conservative.

### Regression Risk
✓ Low. No changes to `DashboardContext.tsx`, `WidgetRegistry.ts`, or `types.ts`. Cloud Function changes preserve existing signatures. Escape-propagation fixes are scoped to the affected menus.

### Missing Tests
✓ Adequate coverage. New regression tests accompany each fix (NextUp overnight/missing-timestamp cases, PopcornBallsView measure+teardown, Escape-propagation for two menus, InviteAcceptance entities, expireActivityWallShares pagination).

### Overall
**Merge readiness:** Ready
All prior review threads are resolved and verified by the author; the bundle is well-tested and low-risk for the `dev-paul` → `main` promotion.

---
_This review was generated automatically. All findings should be verified by a human before merging._

---
_Generated by [Claude Code](https://claude.ai/code)_
